File: market_scanner.py
"""Market scanner for monitoring Polymarket prices."""
import logging
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from py_clob_client.client import ClobClient
from py_clob_client.clob_types import BookParams

logger = logging.getLogger(__name__)


@dataclass
class MarketPrice:
    """Price data for a binary market."""
    token_id: str
    yes_price: float
    no_price: float
    yes_token_id: str
    no_token_id: str
    condition_id: str
    question: str
    total_price: float
    
    @property
    def arbitrage_spread(self) -> float:
        """Calculate the arbitrage spread (1.0 - total_price)."""
        return 1.0 - self.total_price


class MarketScanner:
    """Scans Polymarket for price data."""
    
    def __init__(self, client: ClobClient):
        """Initialize the market scanner.
        
        Args:
            client: Authenticated Polymarket CLOB client
        """
        self.client = client
        self._market_cache: Dict[str, dict] = {}
    
    def get_active_markets(self, limit: int = 100) -> List[dict]:
        """Fetch active binary markets.

        Args:
            limit: Maximum number of markets to return

        Returns:
            List of market data dictionaries
        """
        try:
            response = self.client.get_simplified_markets()

            # Extract markets from response
            if isinstance(response, dict):
                markets = response.get('data', [])
            else:
                markets = response

            # Filter for binary markets (YES/NO) with tradable prices
            binary_markets = []
            for market in markets:
                # Skip if not a dict
                if not isinstance(market, dict):
                    continue

                # Check if market has exactly 2 outcomes (YES/NO)
                tokens = market.get('tokens', [])
                if len(tokens) != 2:
                    continue

                # Check if prices are valid and tradable (not fully resolved)
                # Markets with prices at 0 or 1 are resolved
                p0 = tokens[0].get('price', 0)
                p1 = tokens[1].get('price', 0)

                # Skip fully resolved markets (prices are exactly 0 or 1)
                if (p0 == 0 or p0 == 1) and (p1 == 0 or p1 == 1):
                    continue

                # Skip markets with invalid prices
                if p0 <= 0 or p1 <= 0:
                    continue

                binary_markets.append(market)
                # Cache market metadata
                self._market_cache[market['condition_id']] = market

                if len(binary_markets) >= limit:
                    break

            logger.info(f"Found {len(binary_markets)} active binary markets")
            return binary_markets

        except Exception as e:
            logger.error(f"Error fetching markets: {e}")
            import traceback
            traceback.print_exc()
            return []
    
    def get_market_prices(self, condition_id: str) -> Optional[MarketPrice]:
        """Get current prices for a binary market.

        Args:
            condition_id: The market condition ID

        Returns:
            MarketPrice object or None if error
        """
        try:
            # Get market metadata from cache
            market = self._market_cache.get(condition_id)
            if not market:
                logger.warning(f"Market {condition_id} not in cache")
                return None

            tokens = market.get('tokens', [])
            if len(tokens) != 2:
                return None

            # Get token data for both outcomes
            token_0 = tokens[0]
            token_1 = tokens[1]

            # Get prices directly from cached market data
            price_0 = float(token_0.get('price', 0))
            price_1 = float(token_1.get('price', 0))

            # Skip if prices are invalid
            if price_0 <= 0 or price_1 <= 0:
                logger.debug(f"Invalid prices for {condition_id}: {price_0}, {price_1}")
                return None

            # Create descriptive question from outcomes
            outcome_0 = token_0.get('outcome', 'Option 1')
            outcome_1 = token_1.get('outcome', 'Option 2')
            question = f"{outcome_0} vs {outcome_1}"

            return MarketPrice(
                token_id=condition_id,
                yes_price=price_0,
                no_price=price_1,
                yes_token_id=str(token_0['token_id']),
                no_token_id=str(token_1['token_id']),
                condition_id=condition_id,
                question=question,
                total_price=price_0 + price_1
            )

        except Exception as e:
            logger.error(f"Error getting prices for {condition_id}: {e}")
            return None
    
    def _get_token_price(self, token_id: str, side: str = 'BUY') -> Optional[float]:
        """Get current price for a specific token.
        
        Args:
            token_id: The token ID
            side: BUY or SELL
            
        Returns:
            Price as float or None
        """
        try:
            price = self.client.get_price(token_id, side=side)
            return float(price['price']) if price else None
        except Exception as e:
            logger.debug(f"Error getting price for token {token_id}: {e}")
            return None
    
    def scan_for_arbitrage(self, threshold: float = 0.99) -> List[MarketPrice]:
        """Scan all markets for arbitrage opportunities.
        
        Args:
            threshold: Maximum acceptable total price (YES + NO)
            
        Returns:
            List of markets with arbitrage opportunities
        """
        opportunities = []
        
        # Get active markets
        markets = self.get_active_markets()
        
        for market in markets:
            condition_id = market['condition_id']
            prices = self.get_market_prices(condition_id)
            
            if prices and prices.total_price < threshold:
                logger.info(
                    f"Arbitrage opportunity: {prices.question[:50]}... "
                    f"YES={prices.yes_price:.4f} NO={prices.no_price:.4f} "
                    f"Total={prices.total_price:.4f} Spread={prices.arbitrage_spread:.4f}"
                )
                opportunities.append(prices)
        
        return opportunities

File: src/market_scanner.rs
//! Market scanner for monitoring Polymarket prices.

use anyhow::Result;
use log::{debug, error, info, warn};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketPrice {
    pub token_id: String,
    pub yes_price: f64,
    pub no_price: f64,
    pub yes_token_id: String,
    pub no_token_id: String,
    pub condition_id: String,
    pub question: String,
    pub total_price: f64,
}

impl MarketPrice {
    /// Calculate the arbitrage spread (1.0 - total_price)
    pub fn arbitrage_spread(&self) -> f64 {
        1.0 - self.total_price
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Token {
    token_id: String,
    outcome: String,
    price: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Market {
    condition_id: String,
    tokens: Vec<Token>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct SimplifiedMarketsResponse {
    data: Vec<Market>,
}

pub struct MarketScanner {
    client: reqwest::Client,
    base_url: String,
    market_cache: HashMap<String, Market>,
}

impl MarketScanner {
    /// Initialize the market scanner
    pub fn new(base_url: &str) -> Self {
        Self {
            client: reqwest::Client::new(),
            base_url: base_url.to_string(),
            market_cache: HashMap::new(),
        }
    }

    /// Fetch active binary markets
    pub async fn get_active_markets(&mut self, limit: usize) -> Result<Vec<Market>> {
        let url = format!("{}/simplified-markets", self.base_url);

        let response = self
            .client
            .get(&url)
            .send()
            .await?
            .json::<SimplifiedMarketsResponse>()
            .await?;

        let mut binary_markets = Vec::new();

        for market in response.data {
            // Check if market has exactly 2 outcomes (YES/NO)
            if market.tokens.len() != 2 {
                continue;
            }

            let p0 = market.tokens[0].price;
            let p1 = market.tokens[1].price;

            // Skip fully resolved markets (prices are exactly 0 or 1)
            if (p0 == 0.0 || p0 == 1.0) && (p1 == 0.0 || p1 == 1.0) {
                continue;
            }

            // Skip markets with invalid prices
            if p0 <= 0.0 || p1 <= 0.0 {
                continue;
            }

            // Cache market metadata
            self.market_cache
                .insert(market.condition_id.clone(), market.clone());

            binary_markets.push(market);

            if binary_markets.len() >= limit {
                break;
            }
        }

        info!("Found {} active binary markets", binary_markets.len());
        Ok(binary_markets)
    }

    /// Get current prices for a binary market
    pub fn get_market_prices(&self, condition_id: &str) -> Option<MarketPrice> {
        let market = self.market_cache.get(condition_id)?;

        if market.tokens.len() != 2 {
            return None;
        }

        let token_0 = &market.tokens[0];
        let token_1 = &market.tokens[1];

        let price_0 = token_0.price;
        let price_1 = token_1.price;

        // Skip if prices are invalid
        if price_0 <= 0.0 || price_1 <= 0.0 {
            debug!(
                "Invalid prices for {}: {}, {}",
                condition_id, price_0, price_1
            );
            return None;
        }

        // Create descriptive question from outcomes
        let question = format!("{} vs {}", token_0.outcome, token_1.outcome);

        Some(MarketPrice {
            token_id: condition_id.to_string(),
            yes_price: price_0,
            no_price: price_1,
            yes_token_id: token_0.token_id.clone(),
            no_token_id: token_1.token_id.clone(),
            condition_id: condition_id.to_string(),
            question,
            total_price: price_0 + price_1,
        })
    }

    /// Scan all markets for arbitrage opportunities
    pub async fn scan_for_arbitrage(&mut self, threshold: f64) -> Result<Vec<MarketPrice>> {
        let mut opportunities = Vec::new();

        // Get active markets
        let markets = self.get_active_markets(100).await?;

        for market in markets {
            let condition_id = &market.condition_id;
            if let Some(prices) = self.get_market_prices(condition_id) {
                if prices.total_price < threshold {
                    info!(
                        "Arbitrage opportunity: {}... YES={:.4} NO={:.4} Total={:.4} Spread={:.4}",
                        &prices.question[..prices.question.len().min(50)],
                        prices.yes_price,
                        prices.no_price,
                        prices.total_price,
                        prices.arbitrage_spread()
                    );
                    opportunities.push(prices);
                }
            }
        }

        Ok(opportunities)
    }
}
